slug.go

```diff
@@ -297,6 +279,58 @@ func (p *Packer) packWalkFn(root, src, dst string, tarW *tar.Writer, meta *Meta,
 	}
 }
 
+func (p *Packer) resolveLink(root string, path string, header *tar.Header) (*resolvedSymlink, error) {
+	// Read the symlink file to find the destination.
+	target, err := os.Readlink(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read symlink %q: %w", path, err)
+	}
+
+	// Ensure the target is acceptable per the Packer's configuration.
+	if err := p.checkSymlink(root, path, target); err != nil {
```

Comment:
I'm trying to wrap my mind around what's going on, because I worry about damage to this area of the code. On it's face, this seems like it should be a simple process: recursively follow symlinks, and ensure the ultimate target is either within a root directory, or else is explicitly allowed by config.

Are you saying that checkSymlink is a proxy for checking the validity of a target? And otherwise we need to duplicate that validity check within `resolveLink`?

Reply:
What do you think about codifying the important top level rules with a little restructuring of the symlink case?

e.g.

```go
case fm&os.ModeSymlink != 0:
	resolved, err := p.resolveLink(root, path, header) // recursive

	if err != nil {
		return fmt.Errorf("error resolving symlink: %w", err)
	}

	if resolved.IsInRoot(root) {
		// Simply copy the link if the symlink target is also in the root
		header.Typeflag = tar.TypeSymlink
		header.Linkname = filepath.ToSlash(resolved.absTarget)
		break
	} else if p.dereference || resolved.TargetMatchesAny(p.allowSymlinkTargets) {
		// Dereference this symlink by updating the header with the target file
		// details and set writeBody to true so the body will be written.
		header.Typeflag = tar.TypeReg
		header.ModTime = resolved.info.ModTime()
		header.Mode = int64(resolved.info.Mode().Perm())
		header.Size = resolved.info.Size()
		writeBody = true
	}
```